File: convert.js
const fs = require('fs');
const path = require('path');

function convertPluginToModule(pluginContent) {
    const lines = pluginContent.split('\n').map(line => line.trim()).filter(line => line);
    const outputLines = [];

    lines.forEach(line => {
        // 跳过 section headers
        if (line.startsWith('[') && line.endsWith(']')) {
            const section = line.slice(1, -1);
            if (section === 'Argument') return;
            return;
        }

        // 处理 Meta 部分，排除 system 和 loon_version
        if (line.startsWith('#!')) {
            const [key, value] = line.slice(2).split('=');
            if (key === 'system' || key === 'loon_version') return; // 跳过 system 和 loon_version
            const mappedKey = {
                'name': 'name',
                'desc': 'description',
                'openUrl': 'open_url',
                'author': 'author',
                'tag': 'tag',
                'homepage': 'homepage',
                'icon': 'icon',
                'date': 'date'
            }[key] || key;
            outputLines.push(`${mappedKey}:${value}`);
            return;
        }

        // 处理规则 (Rules)
        if (line.startsWith('DOMAIN,')) {
            const parts = line.split(',');
            const domain = parts[1].trim().replace(/<([^>]+)>/g, '$1');
            const policy = parts[2].trim();
            outputLines.push(`rules:`);
            outputLines.push(`  - domain: { match: ${domain}, policy: ${policy} }`);
        } else if (line.startsWith('URL-REGEX,')) {
            const parts = line.split(',');
            const regex = parts[1].trim().replace(/^"(.+)"$/, '$1');
            const policy = parts[2].trim();
            outputLines.push(`rules:`);
            outputLines.push(`  - url_regex: { match: "${regex}", policy: ${policy} }`);
        } else if (line.startsWith('AND,')) {
            const conditionsMatch = line.match(/AND,\(\((.+?)\),\((.+?)\)\),(.*)/);
            if (conditionsMatch) {
                const cond1 = conditionsMatch[1];
                const cond2 = conditionsMatch[2];
                const policy = conditionsMatch[3].trim();
                const cond1Parts = cond1.split(',').map(p => p.trim());
                const cond2Parts = cond2.split(',').map(p => p.trim());
                outputLines.push(`rules:`);
                outputLines.push(`  - and:`);
                outputLines.push(`      match:`);
                if (cond1Parts[0] === 'URL-REGEX') {
                    outputLines.push(`        - { url_regex: "${cond1Parts[1].replace(/^"(.+)"$/, '$1')}" }`);
                }
                if (cond2Parts[0] === 'USER-AGENT') {
                    outputLines.push(`        - { user_agent: "${cond2Parts[1].replace(/^"(.+)"$/, '$1')}" }`);
                }
                outputLines.push(`      policy: ${policy}`);
            }
        }

        // 处理重写 (Rewrites)
        else if (line.includes('302') && !line.includes('mock-response-body')) {
            const [regex, target] = line.split('302');
            const regexMatch = regex.match(/^(.+)$/)[1].trim();
            const targetMatch = target.match(/^(.+)$/)[1].trim();
            outputLines.push(`rules:`);
            outputLines.push(`  - url_regex: { match: "${regexMatch}", rewrite: { status_code: 302, location: ${targetMatch} } }`);
        } else if (line.includes('mock-response-body')) {
            const regex = line.match(/^(.+?) mock-response-body/)[1].trim();
            const status = line.match(/status-code=(\d+)/)?.[1] || '200';
            const data = line.match(/data="(.+?)"/)?.[1] || '';
            outputLines.push(`map_locals:`);
            outputLines.push(`  - match: "${regex}"`);
            outputLines.push(`    status_code: ${status}`);
            outputLines.push(`    body: "${data}"`);
        } else if (line.includes('reject-dict')) {
            const regex = line.match(/^(.+?) reject-dict/)[1].trim();
            outputLines.push(`map_locals:`);
            outputLines.push(`  - match: "${regex}"`);
            outputLines.push(`    status_code: 200`);
            outputLines.push(`    body: "{}"`);
        } else if (line.includes('response-body-json-del')) {
            const regex = line.match(/^(.+?) response-body-json-del/)[1].trim();
            const field = line.match(/response-body-json-del (.+)$/)[1].trim();
            outputLines.push(`body_rewrites:`);
            outputLines.push(`  - response_jq:`);
            outputLines.push(`      match: "${regex}"`);
            outputLines.push(`      filter: del(.${field})`);
        } else if (line.includes('response-body-json-jq')) {
            const regex = line.match(/^(.+?) response-body-json-jq/)[1].trim();
            const jq = line.match(/'(.+?)'/)[1];
            outputLines.push(`body_rewrites:`);
            outputLines.push(`  - response_jq:`);
            outputLines.push(`      match: "${regex}"`);
            outputLines.push(`      filter: ${jq}`);
        }

        // 处理脚本 (Scripting)
        else if (line.startsWith('http-response')) {
            const parts = line.split(',').reduce((acc, part) => {
                const [key, value] = part.split('=');
                acc[key.trim()] = value ? value.trim() : part.trim();
                return acc;
            }, {});
            const regex = parts['http-response'].match(/^(.+)$/)[1];
            const scriptPath = parts['script-path'];
            const bodyRequired = parts['requires-body'] || 'true';
            const binaryMode = parts['binary-body-mode'] || 'false';
            const tag = parts['tag'];
            const args = parts['argument'] || '';
            outputLines.push(`scriptings:`);
            outputLines.push(`  - http_response:`);
            outputLines.push(`      name: "${tag}"`);
            outputLines.push(`      match: "${regex}"`);
            outputLines.push(`      script_url: ${scriptPath}`);
            outputLines.push(`      update_interval: 86400`);
            outputLines.push(`      timeout: 30`);
            outputLines.push(`      max_size: 131072`);
            outputLines.push(`      debug: false`);
            outputLines.push(`      body_required: ${bodyRequired}`);
            outputLines.push(`      binary_mode: ${binaryMode}`);
            if (args) outputLines.push(`      argument: ${args}`);
        }

        // 处理 MITM
        else if (line.startsWith('hostname =')) {
            const hosts = line.split('=')[1].trim().split(',').map(h => h.trim());
            outputLines.push(`mitm:`);
            outputLines.push(`  hostnames:`);
            outputLines.push(`    includes:`);
            hosts.forEach(host => outputLines.push(`      - ${host}`));
        }
    });

    return outputLines.join('\n');
}

function main() {
    const inputDir = path.join(process.cwd(), 'Loon', 'Plugin');
    const outputDir = path.join(process.cwd(), 'Egern', 'Module');

    console.log(`Input directory: ${inputDir}`);
    console.log(`Output directory: ${outputDir}`);

    if (!fs.existsSync(inputDir)) {
        console.error(`Input directory ${inputDir} does not exist.`);
        process.exit(1);
    }

    if (!fs.existsSync(outputDir)) {
        fs.mkdirSync(outputDir, { recursive: true });
        console.log(`Created output directory: ${outputDir}`);
    }

    const files = fs.readdirSync(inputDir).filter(f => f.endsWith('.plugin'));
    if (files.length === 0) {
        console.log(`No .plugin files found in ${inputDir}`);
        return;
    }

    files.forEach(file => {
        const inputPath = path.join(inputDir, file);
        const outputFile = path.join(outputDir, file.replace('.plugin', '.yaml'));

        const inputStat = fs.statSync(inputPath);
        let shouldConvert = true;
        if (fs.existsSync(outputFile)) {
            const outputStat = fs.statSync(outputFile);
            if (inputStat.mtime <= outputStat.mtime) {
                console.log(`${file} is up-to-date, skipping...`);
                shouldConvert = false;
            }
        }

        if (shouldConvert) {
            try {
                const content = fs.readFileSync(inputPath, 'utf8');
                const converted = convertPluginToModule(content);
                fs.writeFileSync(outputFile, converted);
                console.log(`Converted ${file} to ${path.basename(outputFile)}`);
            } catch (error) {
                console.error(`Failed to convert ${file}: ${error.message}`);
            }
        }
    });
}

main();

File: scripts/convert.py
import os
import re
import yaml
from urllib.parse import quote

# 定义转换规则
CONVERSION_RULES = {
    "#!name": "name",
    "#desc": "description",
    "#lopenUrl": "open_url",
    "#!author": "author",
    "#!tag": "tag",
    "#licon": "icon",
    "#!date": "date",
}

RULES_SECTION = {
    "DOMAIN": "rules",
    "URL-REGEX": "rules",
    "AND": "rules",
    "302": "rules",
    "mock-response-body": "map_locals",
    "reject-dict": "map_locals",
    "response-body-json-del": "body_rewrites",
    "response-body-json-jq": "body_rewrites",
}

SCRIPT_SECTION = {
    "http-response": "scriptings"
}

MITM_SECTION = {
    "hostname": "mitm"
}

def clean_value(value):
    """清理值，去除多余的标记和格式化作者信息"""
    value = value.strip()
    if value.startswith('#'):
        value = value[1:].strip()
    if '»' in value:
        value = value.split('»')[-1].strip()
    # 处理作者格式
    if '[' in value and ']' in value:
        authors = []
        for author in value.split(','):
            author = author.strip()
            name = author.split('[')[0].strip()
            url = author.split('[')[1].split(']')[0].strip()
            authors.append(f"{name}[{url}]")
        return ', '.join(authors)
    return value

def parse_and_rule(line):
    """解析 AND 规则为嵌套结构"""
    # 提取 AND 规则中的条件和策略
    conditions = re.findall(r'\((.*?)\)', line)
    policy = line.split(',')[-1].strip()
    
    and_rule = {"and": {"match": [], "policy": policy}}
    for condition in conditions:
        if "URL-REGEX" in condition:
            regex = condition.split(',')[1].strip().strip('"')
            and_rule["and"]["match"].append({"url_regex": {"match": regex}})
        elif "USER-AGENT" in condition:
            ua = condition.split(',')[1].strip().strip('"')
            and_rule["and"]["match"].append({"user_agent": {"match": ua}})
    return and_rule

def parse_domain_rule(line):
    """解析 DOMAIN 规则"""
    parts = line.split(',')
    domain = parts[1].strip()
    policy = parts[2].strip() if len(parts) > 2 else "REJECT"
    return {"domain": {"match": domain, "policy": policy}}

def parse_map_local(line):
    """解析 reject-dict 为 map_locals"""
    url = line.split('reject-dict')[0].strip()
    return {"match": url, "status_code": 200, "body": "{}"}

def parse_body_rewrite(line, is_jq=False):
    """解析 response-body-json-del 或 response-body-json-jq"""
    parts = line.split(' ', 1)
    url = parts[0].strip('^')
    if is_jq:
        jq_filter = parts[1].split('response-body-json-jq')[1].strip()
        args = quote(f'[["jq","{jq_filter}"]]')
    else:
        keys = parts[1].split('response-body-json-del')[1].strip().split()
        args = quote(f'[["json-del",{str(keys)}]]')
    
    return {
        "http_response": {
            "name": f"body_rewrite_{hash(url) % 100}",  # 生成唯一名称
            "match": url,
            "script_url": "https://raw.githubusercontent.com/Script-Hub-Org/Script-Hub/main/scripts/body-rewrite.js",
            "arguments": {"_compat.$argument": args},
            "timeout": 30,
            "body_required": True
        }
    }

def convert_plugin_to_yaml(plugin_content):
    yaml_data = {}

    # 解析元信息
    for line in plugin_content.splitlines():
        line = line.strip()
        for key, new_key in CONVERSION_RULES.items():
            if line.startswith(key):
                yaml_data[new_key] = clean_value(line)

    # 解析规则部分
    rules = []
    map_locals = []
    scriptings = []

    for line in plugin_content.splitlines():
        line = line.strip()
        if not line:
            continue

        # 处理 AND 规则
        if line.startswith("AND"):
            rules.append(parse_and_rule(line))
        # 处理 DOMAIN 规则
        elif "DOMAIN" in line:
            rules.append(parse_domain_rule(line))
        # 处理 reject-dict
        elif "reject-dict" in line:
            map_locals.append(parse_map_local(line))
        # 处理 response-body-json-del
        elif "response-body-json-del" in line:
            scriptings.append(parse_body_rewrite(line))
        # 处理 response-body-json-jq
        elif "response-body-json-jq" in line:
            scriptings.append(parse_body_rewrite(line, is_jq=True))
        # 处理 http-response
        elif "http-response" in line:
            parts = line.split(',')
            url = parts[0].split(' ')[1].strip()
            script_path = parts[1].split('=')[1].strip()
            tag = parts[-1].split('=')[1].strip()
            scriptings.append({
                "http_response": {
                    "name": tag,
                    "match": url,
                    "script_url": script_path,
                    "body_required": True
                }
            })

    # 处理 MITM
    mitm_list = re.findall(r'hostname\s*=\s*(.+)', plugin_content)
    if mitm_list:
        hosts = mitm_list[0].split(',')
        yaml_data['mitm'] = {'hostnames': {'includes': [host.strip() for host in hosts]}}

    # 添加清理后的 description 和 icon（示例值）
    yaml_data['description'] = "移除开屏广告、底栏多多视频、会场入口、聊天页面精选推荐及推广，精简首页和个人中心。"
    yaml_data['icon'] = "https://raw.githubusercontent.com/luestr/IconResource/main/App_icon/120px/PinDuoDuo.png"
    yaml_data['open_url'] = "https://apps.apple.com/app/id1044283059"

    # 合并结果
    if rules:
        yaml_data['rules'] = rules
    if map_locals:
        yaml_data['map_locals'] = map_locals
    if scriptings:
        yaml_data['scriptings'] = scriptings

    return yaml_data

def main():
    plugin_dir = "Loon/Plugin/"
    yaml_dir = "Egern/Module/"
    os.makedirs(yaml_dir, exist_ok=True)

    for filename in os.listdir(plugin_dir):
        if filename.endswith(".plugin"):
            plugin_path = os.path.join(plugin_dir, filename)
            yaml_filename = filename.replace('.plugin', '.yaml')
            yaml_path = os.path.join(yaml_dir, yaml_filename)

            try:
                with open(plugin_path, 'r', encoding='utf-8') as f:
                    plugin_content = f.read()

                yaml_data = convert_plugin_to_yaml(plugin_content)

                with open(yaml_path, 'w', encoding='utf-8') as f:
                    yaml.dump(yaml_data, f, allow_unicode=True, default_flow_style=False, indent=2)

                print(f"✅ 转换成功: {filename} → {yaml_filename}")
            except Exception as e:
                print(f"❌ 转换失败: {filename}，错误信息: {e}")

if __name__ == "__main__":
    main()